torch/csrc/jit/codegen/cuda/contiguity.cpp

```diff
@@ -8,20 +9,455 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
+OrderedIdInformation::OrderedIdInformation(
+    const std::vector<IterDomain*>& ids,
+    const std::vector<IterDomain*>& root_domain,
+    std::shared_ptr<const ConcretizedBroadcastDomains> concrete_info)
+    : active_ids_(root_domain), concrete_info_(concrete_info) {
+  if (ids.empty() || root_domain.empty()) {
+    return;
+  }
+
+  for (const auto root_i : c10::irange(root_domain.size())) {
+    auto root_id = root_domain[root_i]->as<IterDomain>();
+
+    // Initialize id_to_root_ids to map roots to themselves
+    id_to_root_ids_[root_id] = {root_id};
+
+    // Initialize roots as being made up of correctly ordered transforms.
+    consistently_ordered_ids_.emplace(root_id);
+
+    exclusively_consumes_roots_.emplace(root_id);
+  }
+
+  // Compute consistently_ordered_ids_, id_to_root_ids_, and
+  auto exprs = StmtSort::getExprs(
+      ids[0]->fusion(),
+      {ids.begin(), ids.end()},
+      {root_domain.begin(), root_domain.end()});
+
+  for (auto expr : exprs) {
+    OptInDispatch::handle(expr);
+  }
+}
+
+bool OrderedIdInformation::checkExclusivelyConsumesRoots(IterDomain* id) {
+  TORCH_INTERNAL_ASSERT(
+      std::find(active_ids_.begin(), active_ids_.end(), id) !=
+          active_ids_.end(),
+      "Error replaying transforms in contiguous ID checker, expected ",
+      id->toString(),
+      " to be in the active ID set.");
+
+  auto root_id_it = id_to_root_ids_.find(id);
+  TORCH_INTERNAL_ASSERT(
+      root_id_it != id_to_root_ids_.end(),
+      "Error replaying transforms in contiguous ID checker, couldn't find mapped roots of ",
+      id->toString());
+
+  const auto& root_ids = root_id_it->second;
+
+  // Check all the roots of all other ids, to see if any root_ids in id are also
+  // in them.
+  for (auto other_active_id : active_ids_) {
+    if (other_active_id == id || other_active_id == nullptr) {
+      continue;
+    }
+
+    auto root_id_it = id_to_root_ids_.find(other_active_id);
+    TORCH_INTERNAL_ASSERT(
+        root_id_it != id_to_root_ids_.end(),
+        "Error replaying transforms in contiguous ID checker, couldn't find mapped roots of ",
+        other_active_id->toString());
+
+    const auto& other_root_ids = root_id_it->second;
+
+    for (auto other_root_id : other_root_ids) {
+      if (root_ids.has(other_root_id)) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+void OrderedIdInformation::handle(Merge* merge) {
+  // Find inputs in the active_ids_ vector
+  const auto inner_it =
+      std::find(active_ids_.begin(), active_ids_.end(), merge->inner());
+  const auto outer_it =
+      std::find(active_ids_.begin(), active_ids_.end(), merge->outer());
+
+  // If either aren't in active_ids_ it means the inputs were detected to not be
```

Comment:
I think this question depends on:
https://github.com/csarofeen/pytorch/pull/1990/files#r982323359

I believe it is necessary, and this can happen as we want to support contiguous indexing "through" broadcasting that's not realized. I don't know how often this would happen in practice but I'm pretty confident we can make an example for it.

Reply:
Thanks, I see it now.